homeassistant/components/recorder/util.py

```diff
@@ -0,0 +1,73 @@
+"""SQLAlchemy util functions."""
+from contextlib import contextmanager
+import logging
+import time
+
+from .const import DATA_INSTANCE
+
+_LOGGER = logging.getLogger(__name__)
+
+RETRIES = 3
+QUERY_RETRY_WAIT = 0.1
+
+
+@contextmanager
+def session_scope(*, hass=None, session=None):
+    """Provide a transactional scope around a series of operations."""
+    from sqlalchemy.exc import SQLAlchemyError
+
+    if session is None and hass is not None:
+        session = hass.data[DATA_INSTANCE].get_session()
+
+    if session is None:
+        raise RuntimeError('Session required')
+
+    try:
+        yield session
+        session.commit()
+    except SQLAlchemyError as err:
```

Comment:
I prefer this as a general Exception, since it is important to log any error.

There was an issue aboit invalid codepage TypeError that this also fixed.

All errors will be passed on in any case, but if we log it at least users can fix these type of things